File: src/trading_ai/ui/models/paper_commands.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class PaperOrderSide(str, Enum):
    BUY = "BUY"
    SELL = "SELL"


class PaperOrderType(str, Enum):
    MARKET = "MARKET"
    LIMIT = "LIMIT"


class PaperOrderStatus(str, Enum):
    PENDING = "PENDING"
    ACCEPTED = "ACCEPTED"
    PARTIALLY_FILLED = "PARTIALLY_FILLED"
    FILLED = "FILLED"
    CANCELLED = "CANCELLED"
    REPLACED = "REPLACED"
    REJECTED = "REJECTED"


class GovernedActor(BaseModel):
    user_id: str = Field(min_length=1, max_length=128)
    session_id: str = Field(min_length=1, max_length=128)
    roles: list[str] = Field(default_factory=list)
    permissions: list[str] = Field(default_factory=list)


class PaperOrderSubmitRequest(BaseModel):
    environment: Literal["PAPER", "SIMULATION"] = "PAPER"
    symbol: str = Field(min_length=1, max_length=20)
    instrument_type: Literal["EQUITY", "OPTION"] = "EQUITY"
    side: PaperOrderSide
    order_type: PaperOrderType
    quantity: int = Field(gt=0, le=10000)
    limit_price: float | None = Field(default=None, gt=0)
    estimated_price: float | None = Field(default=None, gt=0)
    option_expiry: str | None = None
    option_strike: float | None = Field(default=None, gt=0)
    option_type: Literal["CALL", "PUT"] | None = None
    reason: str = Field(min_length=5, max_length=500)
    confirmation_token: str = Field(min_length=8, max_length=256)
    idempotency_key: str = Field(min_length=8, max_length=128)
    actor: GovernedActor

    @model_validator(mode="after")
    def validate_order(self):
        self.symbol = self.symbol.strip().upper()
        if self.order_type == PaperOrderType.LIMIT and self.limit_price is None:
            raise ValueError("limit_price is required for LIMIT orders")
        if self.instrument_type == "OPTION":
            if not self.option_expiry or self.option_strike is None or not self.option_type:
                raise ValueError(
                    "option_expiry, option_strike, and option_type are required "
                    "for OPTION orders"
                )
        return self


class PaperOrderCancelRequest(BaseModel):
    environment: Literal["PAPER", "SIMULATION"] = "PAPER"
    reason: str = Field(min_length=5, max_length=500)
    confirmation_token: str = Field(min_length=8, max_length=256)
    idempotency_key: str = Field(min_length=8, max_length=128)
    actor: GovernedActor


class PaperOrderReplaceRequest(BaseModel):
    environment: Literal["PAPER", "SIMULATION"] = "PAPER"
    quantity: int | None = Field(default=None, gt=0, le=10000)
    limit_price: float | None = Field(default=None, gt=0)
    reason: str = Field(min_length=5, max_length=500)
    confirmation_token: str = Field(min_length=8, max_length=256)
    idempotency_key: str = Field(min_length=8, max_length=128)
    actor: GovernedActor

    @model_validator(mode="after")
    def require_change(self):
        if self.quantity is None and self.limit_price is None:
            raise ValueError("quantity or limit_price must be provided")
        return self


class PaperOrderRecord(BaseModel):
    order_id: str
    environment: str
    symbol: str
    instrument_type: str
    side: PaperOrderSide
    order_type: PaperOrderType
    quantity: int
    limit_price: float | None = None
    estimated_price: float | None = None
    option_expiry: str | None = None
    option_strike: float | None = None
    option_type: str | None = None
    status: PaperOrderStatus
    reason: str
    actor_user_id: str
    actor_session_id: str
    idempotency_key: str
    created_at: datetime
    updated_at: datetime
    replaced_by_order_id: str | None = None
    rejection_reasons: list[str] = Field(default_factory=list)


class CommandDecision(BaseModel):
    allowed: bool
    status: str
    action: str
    order: PaperOrderRecord | None = None
    message: str
    policy_reasons: list[str] = Field(default_factory=list)
    idempotent_replay: bool = False


class PaperTradingSummary(BaseModel):
    environment: str = "PAPER"
    mode: str = "GOVERNED_PAPER_ONLY"
    live_trading_enabled: bool = False
    total_orders: int = 0
    open_orders: int = 0
    filled_orders: int = 0
    cancelled_orders: int = 0
    rejected_orders: int = 0
    gross_notional: float = 0.0


class PaperTradingState(BaseModel):
    generated_at: datetime
    summary: PaperTradingSummary
    orders: list[PaperOrderRecord] = Field(default_factory=list)
    permissions_required: list[str] = Field(default_factory=list)
    safety_notices: list[str] = Field(default_factory=list)


class AuditEvent(BaseModel):
    event_id: str
    occurred_at: datetime
    event_type: str
    action: str
    outcome: str
    actor_user_id: str
    actor_session_id: str
    environment: str
    resource_id: str | None = None
    reason: str
    detail: str
    idempotency_key: str

File: src/trading_ai/ui/static/paper_commands.js
(() => {
  const nav = document.querySelector("#nav");
  if (!nav || document.querySelector('[data-view="paper-commands"]')) return;

  const button = document.createElement("button");
  button.dataset.view = "paper-commands";
  button.textContent = "Paper Trading";
  nav.insertBefore(
    button,
    document.querySelector('[data-view="execution"]') || null
  );

  const content = document.querySelector("#content");
  const title = document.querySelector("#pageTitle");
  const notice = document.querySelector("#notice");
  const status = document.querySelector("#status");

  const uid = () =>
    `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}`;

  const actor = () => ({
    user_id: localStorage.getItem("paper.user_id") || "local-operator",
    session_id: localStorage.getItem("paper.session_id") || "local-session",
    roles: ["TRADER"],
    permissions: [
      "paper_orders.view",
      "paper_orders.submit",
      "paper_orders.cancel",
      "paper_orders.replace"
    ]
  });

  const request = async (url, options = {}) => {
    const response = await fetch(url, {
      cache: "no-store",
      headers: {"Content-Type": "application/json"},
      ...options
    });
    const payload = await response.json().catch(() => ({}));
    if (!response.ok) {
      const detail = payload.detail || payload;
      throw new Error(
        typeof detail === "string"
          ? detail
          : detail.message || JSON.stringify(detail)
      );
    }
    return payload;
  };

  const money = value => {
    const number = Number(value);
    return Number.isFinite(number)
      ? number.toLocaleString(undefined, {
          style: "currency",
          currency: "USD"
        })
      : "—";
  };

  const render = async () => {
    document.querySelectorAll("#nav button").forEach(
      item => item.classList.toggle(
        "active",
        item.dataset.view === "paper-commands"
      )
    );
    title.textContent = "Governed Paper Trading";
    notice.textContent =
      "Paper-only commands. Live trading remains disabled by server policy.";
    status.textContent = "LOADING";

    try {
      const state = await request("/api/v1/paper-commands");
      const summary = state.summary;
      status.textContent = summary.mode;

      content.innerHTML = `
        <section class="cards">
          <article class="card"><span>Mode</span><strong>${summary.environment}</strong><small>Live trading disabled</small></article>
          <article class="card"><span>Total Orders</span><strong>${summary.total_orders}</strong><small>${summary.open_orders} open</small></article>
          <article class="card"><span>Cancelled</span><strong>${summary.cancelled_orders}</strong><small>${summary.rejected_orders} rejected</small></article>
          <article class="card"><span>Gross Notional</span><strong>${money(summary.gross_notional)}</strong><small>Paper exposure only</small></article>
        </section>

        <section class="grid">
          <article class="panel">
            <small>Governed Command</small>
            <h2>Submit Paper Order</h2>
            <form id="paperOrderForm" class="stack">
              <div class="toolbar">
                <input name="symbol" placeholder="Symbol" value="AAPL" required>
                <select name="side"><option>BUY</option><option>SELL</option></select>
                <select name="order_type"><option>LIMIT</option><option>MARKET</option></select>
                <input name="quantity" type="number" min="1" max="1000" value="1" required>
                <input name="limit_price" type="number" min="0.01" step="0.01" placeholder="Limit price">
              </div>
              <input name="reason" placeholder="Business reason" value="Paper strategy validation" required>
              <label><input name="confirm" type="checkbox" required> I confirm this is a PAPER order.</label>
              <button type="submit">Submit Paper Order</button>
            </form>
            <div id="paperResult" class="notice">No command submitted.</div>
          </article>

          <article class="panel">
            <small>Safety Governance</small>
            <h2>Active Controls</h2>
            <div class="stack">
              ${state.safety_notices.map(item => `<div class="item PASS"><strong>${item}</strong></div>`).join("")}
              <div class="item PASS"><strong>Maximum quantity: 1,000</strong></div>
              <div class="item PASS"><strong>Maximum notional: $100,000</strong></div>
              <div class="item PASS"><strong>Idempotency required for every command</strong></div>
            </div>
          </article>
        </section>

        <section class="panel">
          <small>Order Lifecycle</small>
          <h2>Paper Orders</h2>
          <div class="table-wrap">
            <table>
              <thead><tr><th>Order</th><th>Symbol</th><th>Side</th><th>Type</th><th>Qty</th><th>Price</th><th>Status</th><th>Actions</th></tr></thead>
              <tbody>
                ${state.orders.length ? state.orders.map(order => `
                  <tr>
                    <td>${order.order_id}</td>
                    <td><strong>${order.symbol}</strong></td>
                    <td>${order.side}</td>
                    <td>${order.order_type}</td>
                    <td>${order.quantity}</td>
                    <td>${money(order.limit_price || order.estimated_price)}</td>
                    <td>${order.status}</td>
                    <td>
                      ${["PENDING","ACCEPTED","PARTIALLY_FILLED"].includes(order.status)
                        ? `<button class="paper-cancel" data-order="${order.order_id}">Cancel</button>
                           <button class="paper-replace" data-order="${order.order_id}" data-qty="${order.quantity}" data-price="${order.limit_price || ""}">Replace</button>`
                        : "—"}
                    </td>
                  </tr>`).join("")
                : `<tr><td colspan="8">No paper orders are available.</td></tr>`}
              </tbody>
            </table>
          </div>
        </section>`;

      document.querySelector("#paperOrderForm").onsubmit = async event => {
        event.preventDefault();
        const form = new FormData(event.currentTarget);
        const orderType = form.get("order_type");
        const payload = {
          environment: "PAPER",
          symbol: form.get("symbol"),
          instrument_type: "EQUITY",
          side: form.get("side"),
          order_type: orderType,
          quantity: Number(form.get("quantity")),
          limit_price: orderType === "LIMIT"
            ? Number(form.get("limit_price"))
            : null,
          estimated_price: orderType === "MARKET"
            ? Number(form.get("limit_price")) || null
            : null,
          reason: form.get("reason"),
          confirmation_token: `CONFIRM-PAPER-${uid()}`,
          idempotency_key: `submit-${uid()}`,
          actor: actor()
        };
        try {
          const decision = await request(
            "/api/v1/paper-commands/orders",
            {method: "POST", body: JSON.stringify(payload)}
          );
          document.querySelector("#paperResult").textContent =
            `${decision.status}: ${decision.message}`;
          await render();
        } catch (error) {
          document.querySelector("#paperResult").textContent =
            `Rejected: ${error.message}`;
        }
      };

      document.querySelectorAll(".paper-cancel").forEach(action => {
        action.onclick = async () => {
          const reason = prompt("Cancellation reason:");
          if (!reason) return;
          await request(
            `/api/v1/paper-commands/orders/${action.dataset.order}/cancel`,
            {
              method: "POST",
              body: JSON.stringify({
                environment: "PAPER",
                reason,
                confirmation_token: `CONFIRM-PAPER-${uid()}`,
                idempotency_key: `cancel-${uid()}`,
                actor: actor()
              })
            }
          );
          await render();
        };
      });

      document.querySelectorAll(".paper-replace").forEach(action => {
        action.onclick = async () => {
          const quantity = Number(prompt(
            "Replacement quantity:",
            action.dataset.qty
          ));
          const limitPrice = Number(prompt(
            "Replacement limit price:",
            action.dataset.price
          ));
          const reason = prompt("Replacement reason:");
          if (!quantity || !limitPrice || !reason) return;
          await request(
            `/api/v1/paper-commands/orders/${action.dataset.order}/replace`,
            {
              method: "POST",
              body: JSON.stringify({
                environment: "PAPER",
                quantity,
                limit_price: limitPrice,
                reason,
                confirmation_token: `CONFIRM-PAPER-${uid()}`,
                idempotency_key: `replace-${uid()}`,
                actor: actor()
              })
            }
          );
          await render();
        };
      });
    } catch (error) {
      status.textContent = "FAILED";
      notice.textContent = `Failed to load paper trading: ${error.message}`;
    }
  };

  button.addEventListener("click", event => {
    event.stopImmediatePropagation();
    const url = new URL(location.href);
    url.searchParams.set("view", "paper-commands");
    history.pushState({}, "", url);
    render();
  });

  if (new URLSearchParams(location.search).get("view") === "paper-commands") {
    render();
  }
})();
